simulation/test_dht_storage.cpp

```diff
@@ -0,0 +1,144 @@
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "test.hpp"
+#include "settings.hpp"
+#include "libtorrent/config.hpp"
+#include "libtorrent/session.hpp"
+#include "libtorrent/kademlia/node.hpp" // for verify_message
+#include "libtorrent/bencode.hpp"
+#include "libtorrent/socket_io.hpp" // for hash_address
+#include "libtorrent/broadcast_socket.hpp" // for supports_ipv6
+#include "libtorrent/performance_counters.hpp" // for counters
+#include "libtorrent/random.hpp"
+#include "libtorrent/ed25519.hpp"
+
+#include "libtorrent/kademlia/dht_storage.hpp"
+#include "libtorrent/kademlia/node_id.hpp"
+#include "libtorrent/kademlia/routing_table.hpp"
+#include "libtorrent/kademlia/item.hpp"
+#include "libtorrent/kademlia/dht_observer.hpp"
+#include "libtorrent/ed25519.hpp"
+
+#include "libtorrent/io_service.hpp"
+#include "libtorrent/deadline_timer.hpp"
+#include "libtorrent/address.hpp"
+#include "libtorrent/time.hpp"
+#include "libtorrent/aux_/time.hpp"
+
+#include "simulator/simulator.hpp"
+#include <functional>
+
+using namespace libtorrent;
+using namespace libtorrent::dht;
+using namespace sim;
+using namespace sim::chrono;
+using namespace sim::asio;
+using sim::simulation;
+using sim::default_config;
+using namespace std::placeholders;
+
+namespace
+{
+	dht_settings test_settings() {
+		dht_settings sett;
+		sett.max_torrents = 2;
+		sett.max_dht_items = 2;
+		sett.item_lifetime = seconds(120 * 60).count();
+		return sett;
+	}
+
+	static sha1_hash to_hash(char const *s) {
+		sha1_hash ret;
+		from_hex(s, 40, (char *) &ret[0]);
+		return ret;
+	}
+}
+
+void timer_tick(boost::shared_ptr<dht_storage_interface> s
```

Comment:
this only tests that the items have expired after 3 hours. please also make it test that they _haven't_ expired after 2 hours. It may even make sense to parameterize this to test a few different timeouts